diesel_codegen_new/src/model.rs

```diff
@@ -1,31 +1,76 @@
 use syn;
 
 use attr::Attr;
-use util::struct_ty;
+use util::{struct_ty, str_value_of_attr_with_name};
 
 pub struct Model {
     pub ty: syn::Ty,
     pub attrs: Vec<Attr>,
     pub name: syn::Ident,
     pub generics: syn::Generics,
+    table_name_from_annotation: Option<syn::Ident>,
 }
 
 impl Model {
-    pub fn from_item(item: &syn::Item) -> Result<Self, String> {
+    pub fn from_item(item: &syn::Item, derived_from: &str) -> Result<Self, String> {
         let fields = match item.body {
-            syn::Body::Enum(..) => return Err("cannot be used with enums".into()),
+            syn::Body::Enum(..) => return Err(format!(
+                "#[derive({})] cannot be used with enums", derived_from)),
             syn::Body::Struct(_, ref fields) => fields,
         };
         let attrs = fields.into_iter().map(Attr::from_struct_field).collect();
         let ty = struct_ty(item.ident.clone(), &item.generics);
         let name = item.ident.clone();
         let generics = item.generics.clone();
+        let table_name_from_annotation = str_value_of_attr_with_name(
+            &item.attrs, "table_name").map(syn::Ident::new);
 
         Ok(Model {
             ty: ty,
             attrs: attrs,
             name: name,
             generics: generics,
+            table_name_from_annotation: table_name_from_annotation,
         })
     }
+
+    pub fn table_name(&self) -> syn::Ident {
+        self.table_name_from_annotation.clone().unwrap_or_else(|| {
+            syn::Ident::new(infer_table_name(self.name.as_ref()))
+        })
+    }
+}
+
+pub fn infer_association_name(name: &str) -> String {
+    let mut result = String::with_capacity(name.len());
+    result.push_str(&name[..1].to_lowercase());
+    for character in name[1..].chars() {
+        if character.is_uppercase() {
+            result.push('_');
+            for lowercase in character.to_lowercase() {
+                result.push(lowercase);
+            }
+        } else {
+            result.push(character);
+        }
+    }
+    result
```

Comment:
This function is just a straight copy paste from the old crate. I'd like to avoid having an inflector beyond the absolute simplest cases. It's been a major source of pain in Rails, as it's brittle, will never get everything right, and is impossible to change without breaking people's apps.